text/0000-non-ascii-idents.md

```diff
@@ -0,0 +1,213 @@
+
+## Remaining ASCII-only names
+
+Only ASCII identifiers are allowed within an external block and in the signature of a function declared `#[no_mangle]`.
+Otherwise an error is reported.
+
+Note: These functions interface with other programming languages
+and these may allow different characters or may not apply normalization to identifiers.
+As this is a niche use-case it is excluded from this RFC.
+A future RFC may lift the restriction.
+
+This RFC keeps out-of-line modules without a `#[path]` attribute ASCII-only.
+The allowed character set for names on crates.io is not changed.
+
+Note: This is to avoid dealing with file systems on different systems *right now*.
+A future RFC may allow non-ASCII characters after the file system issues are resolved.
+
+## Confusable detection
+
+Rust compilers should detect confusingly similar Unicode identifiers and warn the user about it.
+
+Note: This is *not* a mandatory for all Rust compilers as it requires considerable implementation effort and is not related to the core function of the compiler. It rather is a tool to detect accidental misspellings and intentional homograph attacks.
+
+A new `confusable_idents` lint is added to the compiler. The default setting is `warn`.
+
+Note: The confusable detection is set to `warn` instead of `deny` to enable forward compatibility. The list of confusable characters will be extended in the future and programs that were once valid would fail to compile.
+
+The confusable detection algorithm is based on [Unicode® Technical Standard #39 Unicode Security Mechanisms Section 4 Confusable Detection][TR39Confusable]. For every distinct identifier X execute the function `skeleton(X)`. If there exist two distinct identifiers X and Y in the same crate where `skeleton(X) = skeleton(Y)` report it. The compiler uses the same mechanism to check if an identifier is too similar to a keyword.
+
+Note: A fast way to implement this is to compute `skeleton` for each identifier once and place the result in a hashmap as a key. If one tries to insert a key that already exists check if the two identifiers differ from each other. If so report the two confusable identifiers.
+
+## Adjustments to the "bad style" lints
+
+Rust [RFC 0430] establishes naming conventions for Rust ASCII identifiers. The *rustc* compiler includes lints to promote these recommendations.
+
+The following names refer to Unicode character categories:
+
+* `Ll`: Letter, Lowercase
+* `Lu`: Letter, Uppercase
+
+These are the three different naming conventions and how their corresponding lints are specified to accommodate non-ASCII codepoints:
+
+* UpperCamelCase/`non_camel_case_types`: The first codepoint must not be in `Ll`. Underscores are not allowed except as a word separator between two codepoints from neither `Lu` or `Ll`.
+* snake_case/`non_snake_case`: Must not contain `Lu` codepoints.
+* SCREAMING_SNAKE_CASE/`non_upper_case_globals`: Must not contain `Ll` codepoints.
```

Comment:
Besides `Ll` and `Lu`, Unicode has `Lt` (titlecase letters) which are a combination of uppercase and lowercase letters, such as ⟨ǈ⟩. They should be barred from both snake_case and SCREAMING_SNAKE_CASE. UpperCamelCase should bar underscores next to `Lt` too.

Reply:
There's no particular reason to want them but there's no particular reason to bar them too I think. Like Lo, they don't fit neatly in the bicameral distinction, and it's best to just not try to impose rules on them. But we can also treat them the same way we treat Lu. I don't see a need to legislate this personally.

Reply:
A titlecase letter looks and means the same as an uppercase letter followed by a lowercase letter. If “Dz_2” violates the style guide, it makes sense for “ǲ_2” to do so too.

Reply:
>  I don't find this too important 

`[:Lt:]` is indeed general category best left sleeping; while the Latin ones are obviously `[:Lu:][:Ll:]`-like, the Greek ones (which turn into `[:Lu:]` followed by a number of `[:Mn:]` under normalization form D) *can* be used in an ALL CAPS context, see https://en.wikipedia.org/wiki/Iota_subscript#Usage.
Amusingly some fonts (e.g. Source Sans Pro) will show a glyph that resembles a lowercase `ι` *next to* the capital letter for those characters, making them look very titlecasey, but this is nevertheless something that can occur in all-caps text<sup>1</sup>.

> we should probably treat them as Lu-ish.

Agreed; while they should not occur when `Lu` is forbidden (`smol_snake_case`), some of them can legitimately occur in `CAPITAL_SNAKE_CASE` (and obviously they should be allowed in `BactrianCamelCase`).

<sup>1</sup> On the subject of interesting edge cases to capitalization in bicameral scripts, Irish is tricky, e.g. https://en.wiktionary.org/wiki/Appendix:Irish_mutations#Effects_of_eclipsis. As far as I know, Unicode's case folding does not currently deal with that; considering that identifiers tend to be written with slightly odd grammar, often omitting articles or prepositions, this is best left undisturbed for now.